### PR TITLE
feat(plugin-chart-table): rewrite with react-table

### PR DESCRIPTION
**Pull request by ktmud**
🏆 Enhancements

Rewrite Table chart `@superset-ui/plugin-chart-table` with [react-table](https://github.com/tannerlinsley/react-table), which is a **headless** table library already used elsewhere in Superset app.

It's highly customizable and very performant. Unlike `jquery.datatables`, this library does not pre-render all rows when pagination is enabled, therefore there is a noticeable performance improvement when applying on large datasets.

This plugin will eventually replace `@superset-ui/legacy-plugin-chart-table`, and the reusable `DataTable` component may replace other data table, too (including pivot table, time series table, `TableCollection` in CRUD list view, and SQL Lab query results preview). 

Snip20200530_21

This PR merges into a feature branch. Remaining TODOs before we can replace `legacy-plugin-chart-table`:

- [x] Update sort icons
- [ ] Sticky header and fixed scrollable area

### CODE REVIEW

**Comment by mistercrunch** [comment]
OMG, so glad to see this! I can't wait to get rid of `jQuery.datatables`

File: packages/superset-ui-chart-composition/package.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
     "access": "public"
   },
   "dependencies": {
-    "@types/react": "^16.9.34",
+    "@types/react": "^16.9.35",
```
Comment (kristw) [inline packages/superset-ui-chart-composition/package.json:29]: let's do all these bumpings that are not related to the table in separate PR.

File: plugins/plugin-chart-table/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import controlPanel from './controlPanel';
+
+export * from './types';
+
+const metadata = new ChartMetadata({
+  canBeAnnotationTypes: ['EVENT', 'INTERVAL'],
+  description: '',
+  name: t('Table'),
+  thumbnail,
+  useLegacyApi: true,
```
Comment (kristw) [inline plugins/plugin-chart-table/src/index.ts:32]: remove?

Comment (ktmud) [reply plugins/plugin-chart-table/src/index.ts:32]: Wouldn't that affect the query API used in Superset? Haven't started migrating the API yet, just the rendering engine.

Will work on that today.

**Comment by ktmud** [comment]
@kristw This one is ready for re-review. The sticky head now works for the most part. Just needed to enable horizontal scrolls and add some magics to column width so very wide tables can look good.

Another issue is there is this weird scroll bar that runs over the sticky head (material-ui which also uses `position: sticky` has it, too):

image

I'll probably return to the pseudo thead approach like Ant Design and the current jquery.datatables.

Will do all these in a later PR.

**Comment by villebro** [comment]
@ktmud we should probably add proper pagination support to the backend (we're currently missing support for `offset`). I can probably throw that in pretty quickly, is that something that would benefit this work?

**Comment by ktmud** [comment]
> @ktmud we should probably add proper pagination support to the backend (we're currently missing support for `offset`). I can probably throw that in pretty quickly, is that something that would benefit this work?

That sounds like a separate issue but definitely a nice thing to have. I can work on that once the API is ready.

**Comment by ktmud** [comment]
Vercel returns 404, close and reopen to see it fixes it.

File: plugins/plugin-chart-table/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  String = 'string',
+  DateTime = 'datetime',
+}
+
+export interface DataColumnMeta {
+  // `key` is what is called `label` in the input props
+  key: string;
+  // `label` is verbose column name used for rendering
+  label: string;
+  isTime?: boolean;
```
[the remark this answers is not in the document]
Comment (kristw) [reply plugins/plugin-chart-table/src/types.ts:35]: If this is only for formatting, there should not be difference between `'metric'` and `'percent-metric'`? because the formatter can be specified by the field `formatter` ?

File: packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+/**
+ * Load sample data for testing
+ * @param props the original props passed to SuperChart
+ * @param pageSize number of records perpage
+ * @param targetSize the target total number of records
+ */
+function loadData(props: TableChartProps, pageSize = 50, targetSize = 1042): TableChartProps {
+  if (!props.queryData) return props;
+  const dat = props.queryData?.data;
```
Comment (kristw) [inline packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:60]: `dat` => `data`?

File: packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  if (dat && dat.records.length > 0) {
+    while (dat.records.length < targetSize) {
```
Comment (kristw) [inline packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:62]:
```ts
const { records } = data;
const { length } = records;
data.records = [...new Array(targetSize).keys()].map(i => records[i % length]);
```

File: packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx (hunk 4 above)
Comment (ktmud) [reply packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:60]: There is a `data` in the namescope already. Maybe I can rename that to `testData`.

File: packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx (hunk 5 above)
Comment (ktmud) [reply packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:62]: The time complexity becomes `O(n)` instead of `O(log n)`, but I guess we can live with in favor of cleaner code.

Comment (kristw) [reply packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:62]: this is also demo so I would say optional. can live with better perf too.

File: plugins/plugin-chart-table/src/DataTable/Styles.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import styled from '@superset-ui/style';
+
+export default styled.div`
```
[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/DataTable/Styles.tsx:3]: I think it totally makes sense to do nested styles when needed. It makes styling overrides much easier. That's the whole point of *Cascading* Style Sheets. As long as their is not classname conflicts and the components are self-contained (i.e., smaller components can be reused when needed), it should be OK to do cascading styles.

I had some bad experience dealing with `react-select` styling where each component is styled separated and I had to call style functions for each component which is neither efficient nor safer. I think similar frustration was also one of the reasons why Airbnb is moving away from Aphrodite.

File: plugins/plugin-chart-table/src/transformProps.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+import { TableChartProps, TableChartTransformedProps, DataType } from './types';
+
+const { PERCENT_3_POINT } = NumberFormats;
+
+/**
+ * Consolidate list of metrics to string, identified by its unique identifier
+ */
+function consolidateMetricShape(metric: QueryFormDataMetric) {
+  if (typeof metric === 'string') return metric;
+  // even thought `metric.optionName` is more unique, it's not used
```
[the remark this answers is not in the document]
Comment (felixcodes) [reply plugins/plugin-chart-table/src/transformProps.ts:33]: _even though_

[the remark this answers is not in the document]
Comment (kristw) [reply plugins/plugin-chart-table/src/transformProps.ts:33]: lol. thank you native speaker

File: plugins/plugin-chart-table/src/DataTable/Styles.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (kristw) [reply plugins/plugin-chart-table/src/DataTable/Styles.tsx:3]: ic

File: plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+/**
+ * Generate numeric page items in the form of
+ *  first, ..., prev, current, next, ..., last
+ */
+function generatePageItems(pageCount: number, currentPage: number, maxPageItemCount: number) {
+  let pageItems: (string | number)[] = [currentPage];
+  if (maxPageItemCount < 7) {
+    throw new Error('Must allow at least 5 page items');
```
[the remark this answers is not in the document]
Comment (suddjian) [reply plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx:36]: should this be a variable?

[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx:36]: Good call!

File: plugins/plugin-chart-table/src/DataTable/components/SelectPageSize.tsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    );
+  }
+  return (
+    <span className="dt-select-page-size">
+      Show{' '}
+      <FormControl
+        bsSize="small"
+        componentClass="select"
+        value={currentSize}
+        style={{
```
[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/DataTable/components/SelectPageSize.tsx:49]: Fixed this temporarily by adding `form-inline` to the parent `<span>`. Will refactor when removing Bootstrap dependency.

File: plugins/plugin-chart-table/src/DataTable/hooks/useSticky.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+export interface UseStickyTableOptions {
+  getStickyTableSize: () => StickyTableSize | null;
+}
+
+export interface UseStickyInstanceProps extends UseStickyTableOptions {
+  updateStickyTableSize: () => void;
+}
+
+export interface StickyTableSize {
+  tableWidth: number; // scrollable area width
```
[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/DataTable/hooks/useSticky.ts:31]: Wanted to differentiate with `containerWidth` and `columnWidth`, etc, but I guess `width` is good enough since this is a table level hook.

File: plugins/plugin-chart-table/src/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/types.ts:35]: Makes sense, will update.

File: plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  maxPageItemCount?: number;
+  ellipsis?: string; // content for ellipsis item
+  gotoPage: (page: number) => void; // `page` is zero-based
+}
+
+/**
+ * Generate numeric page items in the form of
+ *  first, ..., prev, current, next, ..., last
+ */
+function generatePageItems(pageCount: number, currentPage: number, maxPageItemCount: number) {
```
[the remark this answers is not in the document]
Comment (kristw) [reply plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx:37]: I modify the last part to simplify and add more comments.
Also performs better on jsperf. https://jsperf.com/xdsga

```ts
function generatePageItems(pageCount: number, currentPage: number, maxPageItemCount: number) {
  if (maxPageItemCount < 7) {
    throw new Error('Must allow at least 7 page items');
  }

  if (pageCount <= maxPageItemCount) {
    return [...new Array(pageCount).keys()];
  }

  const rightHalf = (maxPageItemCount % 2 === 1 ? (maxPageItemCount - 1) : (maxPageItemCount)) / 2;
  const leftHalf = maxPageItemCount - rightHalf - 1;
  const draftRightmostPage = Math.min(pageCount - 1, currentPage + rightHalf);
  // if the right side of current page has fewer items than required
  // have to get more items from the left side
  const rightDiff = currentPage + rightHalf - draftRightmostPage;
  const leftmostPage = Math.max(0, currentPage - leftHalf - rightDiff);
  const rightmostPage = leftmostPage + maxPageItemCount - 1;
  const last = pageCount - 1;

  const output = [0, leftmostPage === 0 ? 1 : 'prev-more'];
  for (let i = leftmostPage + 2; i <= rightmostPage - 2; i++) {
    output.push(i);
  }
  output.push(rightmostPage === last ? last - 1 : 'next-more');
  output.push(last);

  return output;
}
```

File: plugins/plugin-chart-table/src/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/types.ts]: I removed these properties, and changed to use `dataType` instead. It now assumes `isMetric` when `dataType === 'number'`, which is a safe assumption at least for now.

File: plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (ktmud) [reply plugins/plugin-chart-table/src/DataTable/components/Pagination.tsx:37]: Found another simpler and [faster](https://jsperf.com/xdsga/4) solution:

```ts
// first, ..., prev, current, next, ..., last
const MINIMAL_PAGE_ITEM_COUNT = 7;

/**
 * Generate numeric page items around current page.
 *   - Always include first and last page
 *   - Add ellipsis if needed
 */
function generatePageItems(total: number, current: number, width: number) {
  if (width < MINIMAL_PAGE_ITEM_COUNT) {
    throw new Error(`Must allow at least ${MINIMAL_PAGE_ITEM_COUNT} page items`);
  }
  if (width % 2 === 0) {
    throw new Error(`Must allow odd number of page items`);
  }
  if (total < width) {
    return [...new Array(total).keys()];
  }
  const left = Math.max(0, Math.min(total - width, current - Math.floor(width / 2)));
  const items: (string | number)[] = new Array(width);
  for (let i = 0; i < width; i += 1) {
    items[i] = i + left;
  }
  // replace non-ending items with placeholders
  if (items[0] > 0) {
    items[0] = 0;
    items[1] = 'prev-more';
  }
  if (items[items.length - 1] < total - 1) {
    items[items.length - 1] = total - 1;
    items[items.length - 2] = 'next-more';
  }
  return items;
}
```
